### ISSUE TITLE
Problems with Milestone Run in RZ

### ISSUE THREAD

**Issue by EZoni** [ECP-WarpX/WarpX#2462]
In preparation of the last milestone report, we ran some simulations of 2 plasma stages in RZ geometry, which did not seem to produce correct results.

Here's the input file that was used: [inputs_rz.txt](https://github.com/ECP-WarpX/WarpX/files/7392075/inputs_rz.txt). The run was performed on Summit using 4 nodes (24 GPUs).

Here's a movie showing the time evolution of the component Ez of the electric field:

https://user-images.githubusercontent.com/59625522/138336562-1299f4cb-a560-4178-8c10-169e17171b79.mp4

Note that this was maybe run on someone's branch instead of the `development` branch. Certainly one would need to rerun this test with the latest version of the code in order to confirm the issue.

**Comment by RemiLehe** [comment]
For comparison, here are the result from a **3D** script, which should in principle simulate the same physics (albeit with different numerical parameters):
![img008500](https://user-images.githubusercontent.com/6685781/138529292-bf44348b-ced2-4f7a-838f-49d37eaaa6de.png)
This snapshot should correspond roughly to iteration 1500 from the above RZ movie. However, the fields significantly different.

Here is the corresponding input script: [input.txt](https://github.com/ECP-WarpX/WarpX/files/7401646/input.txt)

**Comment by EZoni** [comment]
Small update: I re-ran the test with the latest version of WarpX and can confirm that I still see the issue reported here. Note that the RZ test runs quite quickly even on a local desktop computer (e.g. using as little as 8 CPU cores).

**Comment by EZoni** [comment]
@RemiLehe One difference between the 3D input and the RZ input is about the plasma lenses.

In the 3D input we set:
```
particles.B_ext_particle_init_style = repeated_plasma_lens
particles.repeated_plasma_lens_period = 0.35
particles.repeated_plasma_lens_starts = 0.34
particles.repeated_plasma_lens_lengths = 0.002
particles.repeated_plasma_lens_strengths_B = 1369533.545948272
```

While in the RZ input we set:
```
particles.E_ext_particle_init_style = repeated_plasma_lens
particles.B_ext_particle_init_style = repeated_plasma_lens
particles.repeated_plasma_lens_period = 0.35
particles.repeated_plasma_lens_starts = 0.32
particles.repeated_plasma_lens_lengths = 0.03
particles.repeated_plasma_lens_strengths_E = 205291085141280.0
particles.repeated_plasma_lens_strengths_B = 684756.1952742934
```

So just double-checking: are these differences (e.g. one order of magnitude of difference in `repeated_plasma_lens_lengths`, `repeated_plasma_lens_strengths_B` and same for E which is absent in the 3D input) set up on purpose?

**Comment by EZoni** [comment]
This is an updated version of the original input file (with only small changes that do not change the results significantly), sorted so that it can be compared more easily with the input file produced by a PICMI input script: [inputs_rz.txt](https://github.com/ECP-WarpX/WarpX/files/8536690/inputs_rz.txt).

Here's a plot of Ez after 4000 time steps, obtained running on 24 GPUs on Summit, with AMReX (22.04-24-gf2d7f67848de), WarpX (21.10-564-gc8e49fb7ee15) and PICSAR (a621a5a41fb8):
<p align="center">
    <img src="https://user-images.githubusercontent.com/59625522/164593578-cd826dd6-62ec-4af0-81a2-77eb7be88ac5.png" width="600">
</p>

**Comment by EZoni** [comment]
As a result of the latest debugging efforts, these are the most prominent conclusions:
1. Current correction to guarantee charge conservation plays a fundamental role and needs to be activated for these runs.
2. The number of guard cells also plays a relatively significant role, once the current correction is activated as in point 1.
3. For tests similar to the one reported in this issue, we noticed that it is safer if the plasma is contained within the simulation domain (in other words, it is safer if the zmax of the domain strictly contains - does not overlap with - the zmin of the plasma, extending beyond the latter by a few cells).
4. For tests similar to the one reported in this issue, we noticed that it is safer if the laser does not interact with the edge of the plasma in a destructive way (in other words, it is safer if the plasma is extended transversally, so that the laser field amplitude is small at the edge of the plasma).

The following plot shows the results obtained by just applying corrections 1. and 2. above, namely activating the current correction and using 32 guard cells in both the r and z directions, to be compared with the plot in the previous comment:
<p align="center">
    <img src="https://user-images.githubusercontent.com/59625522/167198344-cd8aabab-99ad-4d07-b010-c8b024a9baaa.png" width="600">
</p>

**Comment by EZoni** [comment]
Here's the result for a setup very similar to the original one, [inputs_issue_2462.txt](https://github.com/ECP-WarpX/WarpX/files/8672138/inputs_issue_2462.txt), run on 24 GPUs on Summit, with AMReX (22.04-24-gf2d7f67848de), WarpX (21.10-585-g4a7c47678d42-dirty) and PICSAR (a621a5a41fb8), with the only caveat that an assert on current correction and periodic single box option must be commented out in WarpX (corresponding PR not yet merged).

The videos shows the electric field Ey and Ez (mirrored), the electron beam, and the first and second plasma stages:

https://user-images.githubusercontent.com/59625522/167910167-04c97651-a114-4cc2-a745-b07cd2f6d022.mp4

https://user-images.githubusercontent.com/59625522/167916440-c049d7fb-c777-4938-9126-56d4041ad684.mp4

**Comment by EZoni** [comment]
The following videos show the electric field Ey and Ez (mirrored), the electron beam, and the first and second plasma stages, with a combination of fixes from #3092, #3101, and #3104:

https://user-images.githubusercontent.com/59625522/168146502-23e3ea84-a75f-475b-95df-4363f60593a7.mp4

https://user-images.githubusercontent.com/59625522/168146516-4d89d4cc-b1ff-4a38-bf67-757f51653c0e.mp4

**Comment by RemiLehe** [comment]
This was fixed by a series of recent PRs.
